pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -264,15 +268,42 @@ func (h *upgradeHandler) OnChanged(_ string, upgrade *harvesterv1.Upgrade) (*har
 			return h.upgradeClient.Update(toUpdate)
 		}
 
+		// Upgrade Eligibility Check
 		isEligible, reason := upgradeEligibilityCheck(upgrade, repoInfo)
 
 		if !isEligible {
 			setUpgradeCompletedCondition(toUpdate, StateFailed, corev1.ConditionFalse, reason, "")
 			return h.upgradeClient.Update(toUpdate)
 		}
 
-		logrus.Debug("Start preparing nodes for upgrade")
-		if _, err := h.planClient.Create(preparePlan(upgrade)); err != nil && !apierrors.IsAlreadyExists(err) {
+		// Node Preparation (Image Preloading)
+		imagePreloadStrategy := settings.ImagePreloadStrategy.Get()
+		logrus.WithFields(logrus.Fields{
+			"upgrade_namespace":      upgrade.Namespace,
+			"upgrade_name":           upgrade.Name,
+			"image_preload_strategy": imagePreloadStrategy,
+		}).Info("start preparing nodes for upgrade")
+
+		nodes, err := h.nodeCache.List(labels.Everything())
+		if err != nil {
+			return upgrade, err
+		}
+
+		switch imagePreloadStrategy {
+		case settings.SkipImagePreload:
+			for _, node := range nodes {
+				setNodeUpgradeStatus(toUpdate, node.Name, nodeStateImagesPreloaded, "", "")
+			}
+
+			toUpdate.Labels[upgradeStateLabel] = StatePreparingNodes
+			toUpdate.Status.RepoInfo = repoInfoStr
+			setNodesPreparedCondition(toUpdate, corev1.ConditionTrue, "", "")
+			return h.upgradeClient.Update(toUpdate)
+		case settings.ParallelImagePreload:
+			imagePreloadConcurrency = len(nodes)
```

Comment:
Recall @bk201 ever mentioned, in parallel mode, the most challenge is that those nodes will get the ISO from the repo at the same time, that will be a pressure to the underlayer LH, and the CPU & network to the node where repo is running.

Did you have a test on multi-node cluster? Any plan to set a number like max[3, node-count] ?

The other solution is to set a max parallel accessing on repo nginx.

Reply:
No, I haven't tried that in an environment with a considerable number of container images that need to be downloaded since the script simply skip preloading images in same-version upgrades.

We can enforce rate limiting on either side, with `curl --rate-limit` on the clients or once and for all, with Nginx's `limit_rate` directive on the server.

Ideally, clients will compete for their max bandwidth, so from the server's POV, the total uplink bandwidth is almost the same if no rate limitation is applied. The real problem, I guess, is when the client on the same node as the server starts to download and import images. Heavy read and write requests will happen simultaneously on that node, potentially causing instability.

With that said, the `parallel` option is not recommended in virtual environments. We'll have that noted in the setting's description and documents.

Reply:
perhaps we should add another setting to configure the number of parallel job.

Reply:
> the most challenge is that those nodes will get the ISO from the repo at the same time

It was about the image preloading behavior on each node (generating a large amount of I/Os). But the issue only happened in a virtualized environment (all Harvester nodes are VMs). Now we skip unpacking, so likely the preloading will be smooth on bare-metal machines.